### DISCUSSION TITLE
Any one using this with a Joroto X4S?

### DISCUSSION THREAD

**Discussion by MitjaHenner** [doudar/SmartSpin2k#497] [General]
I have a X4S and am thinking about building a SS2K (probably a non PCB version because I should have all the parts from tinkering with 3d printers, lasers and cncs).

I think Bluetooth only shows cadence, but the display shows resistance (maybe it's possible so get something from the sensors directly).

**Comment by doudar** [comment]
We've made a ?couple? for the Joroto S2. Adaptors can be found here https://github.com/doudar/SmartSpin2k/tree/develop/Hardware/Common%20Assets/Arm
I would expect that if it has bluetooth, it would send power, cadence and possibly resistance. If it sends resistance and cadence only, it could be computed on the SS2k, we would just need to add another decoder, which could be fun.

**Reply by MitjaHenner** [reply]
I think the Knob is the same as the X2, but the mount is different, but that should be no problem.
Yea, it's dumb that they only send cadence, when they clearly have all the data.

Is there a bluetooth app for Android that you recommend?

**Reply by doudar** [reply]
I use NRF connect for investigating data.

**Reply by MitjaHenner** [reply]
Thanks, exactly what I was looking for.

**Reply by doudar** [reply]
Let me know what you find out!

Decoders are in https://github.com/doudar/SmartSpin2k/tree/develop/lib/SS2K/src/sensors

Then if it's a custom characteristic, the uuid will need to be added to the scanning and connection routines in BLE_client.cpp

**Reply by doudar** [reply]
I created a discussion here as well: https://www.facebook.com/groups/149984563348738/

**Reply by cagnulein** [reply]
I'm here too :)

**Reply by doudar** [reply]
@MitjaHenner , if you have access to power meter pedals, we (Roberto and/or I) would love to get access to the data to create a power lookup table.

**Reply by MitjaHenner** [reply]
Probably have some more time over Christmas and might be able to borrow a pair.

**Comment by MitjaHenner** [comment]
Looks like it's only reporting the battery level and cadence over bluetooth.

The resistance sensor is just a sliding pot, so it should be pretty easy to grab the % value.

![image](https://github.com/doudar/SmartSpin2k/assets/3903496/9ff49590-5544-44f9-8ed4-fc7d221870f9)

**Comment by jrbarrick** [comment]
I've got the X4S as well, and just received a batch of PCBs.  Can confirm it only reports speed/cadence over BLE, wish I'd actually researched any of this before I got it a couple months ago.  I was able to wire up the pot to the unused GPIO35 and write a bit of code to read the value and calculate the resistance to match what the panel is showing.  The onboard computer sends 3.3V to the pot, so I can run the SS2k inline with the computer connected and only need to connect to ground and the wiper.  I stuck a pullup resistor on GPIO35 as well so I can use the 4095 value as a bit of crude detector to see if the pot is connected or not, as even at full resistance it doesn't quite hit that.

However, it's been about 20 years since I last touched any C++, so all I've done is just drop the few lines of code in the 2 sec maintenance loop so I can print the values over serial as a test.  I've got no idea from here about to actually make this data useful anywhere.

**Reply by doudar** [reply]
I'd love to see a pull request on what you've done.

Basically, as far as reporting power, you need to come up with a lookup table or simple math (what the IC4 uses, essentially cadence * resistance value), and then do:
```
 rtConfig->watts.setSimulate(true);  
//fancy math or your table implementation here
rtConfig->watts.setValue(value.toInt());
```

**Reply by MitjaHenner** [reply]
Did you get it to work? I hope I finally have time this year to build a SS2k and would probably do something similar and connect the sensor and pot directly to the esp, so I don't need the bulky battery powered box any more.

**Reply by jrbarrick** [reply]
> Did you get it to work? I hope I finally have time this year to build a SS2k and would probably do something similar and connect the sensor and pot directly to the esp, so I don't need the bulky battery powered box any more.

Did I get it to work? Technically...
Is it one of the biggest hackjobs I've ever done? Definitely.

Plus side, I was able to connect to the pot directly and pull the resistance, and even somewhat reverse engineer the power calcs so I could usually get within a few % of the watts on the display.  I cut the traces on the RSIN and RSOUT pins on the SS2k, used them for the pot (connected to unused analog input) and ground, with a pullup resistor on the input for when the bike is off/not connected.  Picture here:
https://imgur.com/a/S9neZCv

Down side, I could never get the SS2K to connect to the bike itself in any useful way, I think because it only reports as a cadence sensor and not a power meter, so it wouldn't use it for power calcs.  The bike is also a BLE receiver for HR data, and it would try connecting to the SS2K for HR data and mostly just throw errors.

At the same time, I was also messing around with my own DIY power meter that I found here:
https://imateapot.dev/homemade-power-meter/
https://gitlab.com/tbressers/power/-/wikis/home

I refactored the tbressers code into C++ from the ino files, and got the device built and hooked up, but the actual power reported never seemed all the accurate.  However, now that I had a power meter that SS2K was happy with, I could replace its power output with what I calculated from the Joroto instead.  That actually worked really well, and I was able to do a full 12 week Zwift training plan with virtually no issues, other than the DIY power meter not liking a cadences above 100 (fine with me, I'm not a serious rider).

I have a Joroto branch here with all the changes I made:
https://github.com/jrbarrick/SmartSpin2k/tree/joroto

This is the code I used to calculate the power and replace the power meter output in SensorCollector.cpp:
```cpp
if (sensorData->hasPower() && !rtConfig->watts.getSimulate()) {
    if ((charUUID == PELOTON_DATA_UUID) && !((String(userConfig->getConnectedPowerMeter()) == "none") || (String(userConfig->getConnectedPowerMeter()) == "any"))) {
      // Peloton connected but using BLE Power Meter. So skip power for Peloton UUID.
    } else {
      /*
      // Original code
      int power = sensorData->getPower() * userConfig->getPowerCorrectionFactor();
      rtConfig->watts.setValue(power);
      spinBLEClient.connectedPM = true;
      logBufLength += snprintf(logBuf + logBufLength, kLogBufMaxLength - logBufLength, " PW(%d)", power % 10000);
      */

      // Joroto testing
      int jorotoPower = 0;
      int jorotoCadence = 0;
      int potPercent = 0;
      int potValue = 0;
      
      // Do 2 reads to smooth things a bit
      potValue = analogRead(POT_PIN);
      potValue += analogRead(POT_PIN);
      potValue /= 2;
      if (potValue == 4095) {
        logBufLength += snprintf(logBuf + logBufLength, kLogBufMaxLength - logBufLength, "Pot not connected.");
      }
      //else if (potValue == 0) {
        //logBufLength += snprintf(logBuf + logBufLength, kLogBufMaxLength - logBufLength, "Pot too low.");
      //}
      else {
        // Potentiometer < 50 is min according to Joroto display
        if (potValue < 50) { potPercent = 1; }
        // Potentiometer >=2500 is max according to Joroto display
        else if(potValue >= 2500) { potPercent = 100; }
        // Percentage on Joroto display is the potentiometer value / 25
        else { potPercent = potValue / 25; }
        rtConfig->setMinResistance(MIN_PELOTON_RESISTANCE);
        rtConfig->setMaxResistance(MAX_PELOTON_RESISTANCE);
        rtConfig->resistance.setValue(potPercent);
        logBufLength += snprintf(logBuf + logBufLength, kLogBufMaxLength - logBufLength, " RS(%d)", potPercent);
        jorotoCadence = rtConfig->cad.getValue();
        // Cadence < 15 is treated as 0 by the Joroto display
        if (jorotoCadence < 15) { jorotoCadence = 0; }
        // Joroto power calc gets weird when resistance is below 10%
        if (potPercent < 10) {
          jorotoPower = potValue / 25.0f * pow((jorotoCadence / 100.0f), 1.5f) * 7.228958f + jorotoCadence - (jorotoCadence / 100.0f) * 60.0f;
          logBufLength += snprintf(logBuf + logBufLength, kLogBufMaxLength - logBufLength, " Low Calc");
        } else { 
          jorotoPower = potValue / 25.0f * pow((jorotoCadence / 100.0f), 1.5f) * 7.228958f + jorotoCadence - 40;
          logBufLength += snprintf(logBuf + logBufLength, kLogBufMaxLength - logBufLength, " Reg Calc");
        }
        // Can sometimes get weird power cals at very low cadence/resistance levels, just set power to 0
        if (jorotoPower < 0) { jorotoPower = 0; }
        logBufLength += snprintf(logBuf + logBufLength, kLogBufMaxLength - logBufLength, " PW(%d)", jorotoPower);
        rtConfig->watts.setValue(jorotoPower);
        spinBLEClient.connectedPM = true;
      } 
    }
  }
```

Here's a screenshot of my spreadsheet of displayed vs calculated power output for a range of resistances/cadences:
https://imgur.com/a/TNmj2pi

I hoping to revisit this sometime in the next few months and see if I can workaround the power meter requirement in SS2K and get it to play nicely.  Also planning to take a look at qdomyos-zwift and see if that won't make things a bit easier, since it seems to be a bit more flexible about the devices it connects to, and I think I could add my power calcs to it and get it to report power back to SS2K.

**Reply by doudar** [reply]
Check out #619 . I've getting close to supporting standalone CSC. Currently it will be subscribed to if a PM device has the characteristic. My plan is to implement another UI option for CSC connection, and then another option to use a saved PowerTable as a lookup table for cadence x (homed)stepper position = watts.

**Reply by jrbarrick** [reply]
Finally had a chance to revisit this a year later.  Had some weirdness with the SS2K refusing to connect to the Joroto bike at first, so I spent a few hours on a wild goose chase until it just started working out of nowhere and has been completely fine ever since.  No idea what was up with that.  Also ran into a firmware bug that causes reboot loops on the Joroto control panel, but turns out that can be fixed by updating the firmware on the panel using the iConsole+ app.

Once the CSC connection was working, it wasn't too hard to kludge up a new sensor that combined the CSC and CyclePower sensors into a Joroto specific one using the code above.  I've only spent a few minutes riding, but seems like everything is working perfectly.  Much cleaner now without having to use the DIY power meter to get cadence. 

https://github.com/jrbarrick/SmartSpin2k/tree/joroto

Note it still requires the hardware modifications I mentioned earlier to connect directly to the potentiometer, so not something that will ever be prime time, but it was a fun little project.
